### PR TITLE
Arihan/mar 8 ability to comment by line of code in pr

### PR DESCRIPTION
**Pull request by arihanv**
- This PR introduces the ability for Maige to add comments inline on changed files. This is done through parsing the git diff file and using the codeComment function.
- The reviewer agent automatically reviews PRs when they are opened and when new commits are pushed to them.
- Also, the reviewer agent can both answer questions about a PR (ex: What is page.tsx about?) and review the changes in files.

### CODE REVIEW

File: lib/agents/reviewer.tsx [hunk 1]
```diff
@@ -14,34 +16,88 @@ const model = new ChatOpenAI({
...
+
+		const result = await executor.call({input})
+		const {output} = result
+
+		return output
+	} else {
+		const prefix = `
+		You are reviewing code snippets from a pull request. Your goal is to provide feedback on the code snippet
+		through using the codeComment function. You must specify the line number for the line you wish to comment on. If there is no reason
+		to comment on a line then don't do anything.
```
Comment (tedspare) [inline lib/agents/reviewer.tsx:64]: 100x prompt engineering

File: lib/tools/prComment.ts [hunk 2]
```diff
@@ -8,22 +8,18 @@ import {addComment} from '~/utils/github'
 export function prComment({octokit, pullId}: {octokit: any; pullId: string}) {
 	return new DynamicStructuredTool({
 		description: 'Adds a comment to a PR',
-		func: async ({comment, severity}) => {
+		func: async ({comment}) => {
 			const footer = `By [Maige](https://maige.app). How's my driving?`
 			const res = await addComment({
 				octokit,
 				issueId: pullId,
-				comment: `${comment}\nThis is a Level ${severity} problem.\n\n${footer}`
```
Comment (tedspare) [inline lib/tools/prComment.ts:16]: Was severity meant to be removed here?

File: lib/agents/reviewer.tsx [hunk 3]
```diff
@@ -1,7 +1,9 @@
 import {initializeAgentExecutorWithOptions} from 'langchain/agents'
 import {ChatOpenAI} from 'langchain/chat_models/openai'
 import {SerpAPI} from 'langchain/tools'
+import parse, {Change, Chunk, File} from 'parse-diff'
 import env from '~/env.mjs'
+import {codeComment} from '~/tools/codeComment'
```
[the remark this answers is not in the document]
Comment (tedspare) [reply lib/agents/reviewer.tsx:6]: YO THIS REVIEWING FEATURE IS SICK

@arihanv flagging this (as a much higher-level PR review than usual, which is SICK):

Since `codeComment` is used below, how could we prevent this kind of error? Eg. prompt engineering ("this snippet is not the full file and might be missing parts etc") or passing longer code snippets?

File: app/api/webhook/github/route.ts [hunk 4]
```diff
@@ -167,6 +167,39 @@ export const POST = async (req: Request) => {
...
+			pull_request: {diff_url: diffUrl}
+		} = payload
+
+		// Get GitHub app instance access token
+		const app = new App({
+			appId: process.env.GITHUB_APP_ID || '',
+			privateKey: process.env.GITHUB_PRIVATE_KEY || ''
+		})
+
+		const octokit = await app.getInstallationOctokit(instanceId)
```
[the remark this answers is not in the document]
Comment (tedspare) [reply app/api/webhook/github/route.ts:181]: Same here. `instanceId` is extracted above.

File: lib/agents/reviewer.tsx [hunk 5]
```diff
@@ -14,34 +16,88 @@ const model = new ChatOpenAI({
...
-			prefix
-		}
-	})
-
-	const result = await executor.call({input})
-	const {output} = result
-
-	return output
+	if (pullId) {
+		const tools = [new SerpAPI(), prComment({octokit, pullId})]
```
[the remark this answers is not in the document]
Comment (tedspare) [reply lib/agents/reviewer.tsx:34]: @arihanv I think this one is wrong?

File: app/api/webhook/github/route.ts [hunk 6]
```diff
@@ -167,6 +167,39 @@ export const POST = async (req: Request) => {
...
+
+		const response = await fetch(diffUrl)
+
+		if (!response.ok) return new Response('Could not fetch diff', {status: 401})
```
[the remark this answers is not in the document]
Comment (tedspare) [reply app/api/webhook/github/route.ts:185]: @arihanv +1

File: lib/tools/prComment.ts (hunk 2 above)
Comment (arihanv) [reply lib/tools/prComment.ts:16]: Yeah

File: lib/tools/prComment.ts [hunk 7]
```diff
@@ -8,22 +8,18 @@ import {addComment} from '~/utils/github'
 export function prComment({octokit, pullId}: {octokit: any; pullId: string}) {
 	return new DynamicStructuredTool({
 		description: 'Adds a comment to a PR',
-		func: async ({comment, severity}) => {
+		func: async ({comment}) => {
```
[the remark this answers is not in the document]
Comment (tedspare) [reply lib/tools/prComment.ts:11]: I said the same, lol

**Comment by arihanv** [comment]
We need to call Maige when the PR is opened. This is because the payload given by PR opened vs PR comments is different. PR opened payload gives us the SHA for the last commit which is needed for the inline code review comments. Doing it with PR comments is much more complicated and requires more API calls to Github. Additionally, it would allow the user to call the review repeatedly on the same PR which doesn't make sense if nothing changed. Thus, I believe the way it is right now is prolly best atm. Let me know what you think :)

**Comment by tedspare** [comment]
maige could you review this?

**Comment by tedspare** [comment]
maige could you review this?

**Comment by tedspare** [comment]
maige could you review this?

**Comment by tedspare** [comment]
maige could you review this?

**Comment by tedspare** [comment]
maige could you review this?

**Comment by tedspare** [comment]
maige could you review this?

**Comment by tedspare** [comment]
maige could you summarize this?

**Comment by tedspare** [comment]
maige could you summarize this?

**Comment by tedspare** [comment]
maige could you summarize this?

**Comment by tedspare** [comment]
maige could you summarize this?

File: app/api/webhook/github/route.ts [hunk 8]
```diff
@@ -167,60 +167,23 @@ export const POST = async (req: Request) => {
...
-
-		return new Response('Reviewed PR', {status: 200})
-	}
-
 	if (
 		!(
 			(action === 'opened' && payload?.issue) ||
-			(action === 'created' && payload?.comment)
+			(action === 'created' && payload?.comment) ||
+			(action === 'opened' && payload?.pull_request)
```
Comment (tedspare) [inline app/api/webhook/github/route.ts:174]: Does this condition prevents [the one below](https://github.com/RubricLab/maige/pull/55/files#diff-52032ba7797e13b0c74f7bc549029f17b98b3143853ec980000d84ce121e38b5R255)?

**Comment by tedspare** [comment]
maige test

**Comment by tedspare** [comment]
maige test

**Review by tedspare — COMMENTED** [review]
Maige test code review comment

File: app/api/webhook/github/route.ts (hunk 8 above)
Comment (tedspare) [reply app/api/webhook/github/route.ts:174]: maige inline code reply

**Review by tedspare — APPROVED** [review]
LGTM! Tested a ton on my end.

**Comment by tedspare** [comment]
maige test
